### PR TITLE
Shared WebSocket

### PR DESCRIPTION
**Pull request by tothandras**
[design doc](https://docs.google.com/document/d/1bveqKsn0J5G68cyMMNzgDhuikgIQ34uVEBa3XR9MJso/edit)

- [x] specification; indexeddb stores, processUrl
- [x] tests (116 unit tests)
- [x] rebase
- [x] update bower package (v1.0.10) for @shanzi
- [x] use guanlecoja

feedback from @shanzi:
- [x] it throws IndexedDB AbortError exception when navigating to a child state "quickly"
- [x] Safari having problem (DOM IDBDatabase Exception 8)
- there are still some issues, probably [IndexedDBShim](https://github.com/axemclion/IndexedDBShim) is needed
- I have opened an [issue](https://github.com/dfahlander/Dexie.js/issues/120) on Dexie, hopefully someone can help me
- [x] support subscribe: false
- [x] control

Will do in other PRs:
- mock
- migrate bower package to (https://github.com/buildbot/buildbot-data-js)
- migrate base to the new service
- travis: run tests, update bower repo
- docs

### CODE REVIEW

**Comment by shanzi** [comment]
😃

**Comment by shanzi** [comment]
Sync state of everything across tabs seems to be too complex.

I am wondering if it is acceptable that different tabs sharing the same web socket connection and cache but not requiring all the tabs automatically update itself to conforms to latest state.

**Comment by sa2ajj** [comment]
a rebase is due.

**Comment by tothandras** [comment]
Please collaborate: [google docs](https://docs.google.com/document/d/1bveqKsn0J5G68cyMMNzgDhuikgIQ34uVEBa3XR9MJso/edit?usp=sharing)

**Comment by tothandras** [comment]
This is the first fully working version, with a few limitations. 
Please review @tardyp, @shanzi!

File: www/data_module/gulpfile.coffee [hunk 1]
```diff
@@ -51,9 +51,9 @@ gulp.task 'build', ['clean'], ->
     gulp.src files.coffee
```
Comment (tardyp) [inline www/data_module/gulpfile.coffee]: Please use guanlecoja for building.

guanlecoja-ui has exactly the same build needs, so please use the same methodology.

File: www/data_module/karma.conf.coffee [hunk 2]
```diff
@@ -1,54 +1,56 @@
...
-    preprocessors: '**/*.coffee': ['ng-classify', 'coffee']
-
-    ngClassifyPreprocessor:
-      options:
-        # any options supported by ng-classify
-        # https://github.com/CaryLandholt/ng-classify#api
-        appName: 'bbData'
-        provider:
-          suffix: 'Service'
+    config.set
```
Comment (tardyp) [inline www/data_module/karma.conf.coffee]: guanlecoja allows to factorize all this boilerplate.

File: www/data_module/src/data.config.coffee [hunk 3]
```diff
@@ -1,5 +1,10 @@
 class LoggingInterceptor extends Config
```
Comment (tardyp) [inline www/data_module/src/data.config.coffee]: probably need to change the name of class, as it does not only do the logging interception now.

File: www/data_module/src/services/data/collection/collection.service.coffee [hunk 4]
```diff
@@ -0,0 +1,86 @@
+class Collection extends Factory
+    constructor: ($q, $injector, $log, dataUtilsService, tabexService, indexedDBService, SPECIFICATION) ->
+        return class CollectionInstance extends Array
```
Comment (tardyp) [inline www/data_module/src/services/data/collection/collection.service.coffee]: extends Array: good idea

File: www/data_module/src/services/indexeddb/indexeddb.service.coffee [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+                    property = query?.property
+                    promise = @properties(promise, property)
+
+                    # 5. fields
+                    fields = query?.field
+                    promise = @fields(promise, fields)
+
+                    return promise
+
+            filter: (promise, filters) ->
```
Comment (tardyp) [inline www/data_module/src/services/indexeddb/indexeddb.service.coffee]: this does not look right to have a function which takes a promise, and do stuff on its results

**Comment by tardyp** [comment]
Now that the design is complete, we should translate the google docs into a rst file, and put it in buildbot docs.

Also indeed, some tests are needed.
It's a little bit big, so hard to review.

File: www/data_module/src/services/indexeddb/indexeddb.service.coffee (hunk 5 above)
Comment (tothandras) [inline www/data_module/src/services/indexeddb/indexeddb.service.coffee]: you are right, I fixed it

**Comment by sa2ajj** [comment]
And one more rebase is necessary :/

**Comment by tothandras** [comment]
Please review @tardyp, @shanzi

**Comment by tardyp** [comment]
LGTM. not enough bw to check why travis fails. @shanzi, please review as well.

**Comment by shanzi** [comment]
@tardyp @tothandras There is also some problem when using the latest `dataService` in new UI. Please have a look at #1810

**Comment by tothandras** [comment]
Thank you for the feedback, I will look at them!

**Comment by tardyp** [comment]
do we still need work on this? I think we should merge it, as it is already big enough.

**Comment by tothandras** [comment]
This could be merged. I am solving the problems that @shanzi noticed. Those
can be in a different PR.

**Pull request merged**